Support COM objects with dynamic keyword

elinor-fung opened the pull request:
This is a port of the support (which is in Framework and the IronLanguages DLR) for using the dynamic keyword with COM objects.

The actual support is in `Microsoft.CSharp`. The tests are under coreclr, since that is set up to test against a native COM server. I did a couple passes to make the ported sources fit in the runtime repo (I'm sure there's more that could be done).

The vast majority of this is a simple copy of the previous support. The commits in this PR are:
- Straight copy of all the sources (first commit [21a9e9a](https://github.com/dotnet/runtime/pull/33060/commits/21a9e9a754e2453bb90a1e8a362e04460b219a28))
- Changes (next 15 commits) to get it building and working
- New tests
- Additional clean-up/consolidation to make it fit a bit better in this repo.

cc @jaredpar @cston 

Resolves #12587

elinor-fung committed:
Remove netfx files. Remove netcoreapp suffix from file names.

elinor-fung committed:
Use SR.Format

elinor-fung committed:
Use built-in type aliases. Suppress warning for generated code.

elinor-fung committed:
AstUtils.Convert -> ComInterop.Helpers.Convert

elinor-fung committed:
AstUtils.Constant -> Expression.Constant

elinor-fung committed:
Fix style violations

elinor-fung committed:
License, style, auto-format

elinor-fung committed:
Set ENABLECOMBINDER. Fix usage of ComBinder in CSharp*Binder

elinor-fung committed:
Copy ComInterop from IronLanguages/dlr. Add to Microsoft.CSharp

elinor-fung committed:
Update namespaces

elinor-fung committed:
TypeUtils.IsComObject -> Marshal.IsComObject

elinor-fung committed:
Use Array.Empty

elinor-fung committed:
Add System.Reflection.* references and allow unsafe

elinor-fung committed:
Remove usage of ContractUtils and System.Security.Permissions

elinor-fung committed:
Make classes internal

elinor-fung committed:
Move error strings to resx. Remove unused strings

elinor-fung committed:
Add test project

Collection tests

elinor-fung committed:
Basic tests

elinor-fung committed:
Event test

elinor-fung committed:
Fix using COM events

elinor-fung committed:
Parameters test

elinor-fung committed:
Move System.Runtime.InteropServices.Variant to shared directory

Reconcile Variant used for dynamic support with shared source

elinor-fung committed:
Use nameof

elinor-fung committed:
Use Unsafe.AsPointer instead of creating dynamic assembly for method

elinor-fung committed:
Move ComEventsMethod and ComEventsSink to shared directory

elinor-fung committed:
Remove warning suppressions

elinor-fung committed:
Remove unused code

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
```
Why is this necessary?  Is this file built into some project other than Corelib and that hasn't been annotated yet for nullable reference types? (Same question for the other occurrences in other files.)

elinor-fung replied:
`Microsoft.CSharp` has not been

stephentoub commented on src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs:
```diff
@@ -170,9 +174,11 @@ public unsafe void CopyFromIndirect(object value)
                     *(uint*)this._typeUnion._unionTypes._byref = (uint)value;
                     break;
 
+#pragma warning disable 618
```
What required this be added?  IIRC, 0618 is for using an obsoleted member.

elinor-fung replied:
`ErrorWrapper` is obsolete. I'll add a comment. (Same with the other `*Wrapper` classes for COM).

stephentoub replied:
> ErrorWrapper is obsolete

Thanks.  My question was really, why wasn't this suppression necessary before?

elinor-fung replied:
Ahh, it looks like this is because those wrapper classes are marked obsolete in the ref assembly for System.Runtime.InteropServices, but not the implementation. So when this file was only used by CoreLib and not going through the ref assembly, it was not considered obsolete.

stephentoub replied:
I see. Ok, thanks.

stephentoub commented on src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml:
```diff
@@ -4,5 +4,7 @@
     <!-- Required by the Expression Evaluator -->
     <type fullname="Microsoft.CSharp.RuntimeBinder.DynamicMetaObjectProviderDebugView/DynamicDebugViewEmptyException"/>
+    <!-- Required for COM event dispatch -->
+    <type fullname="System.Runtime.InteropServices.ComEventsSink"/>
```
This prevents ComEventsSink in Microsoft.CSharp.dll from being trimmed out.  Is it being accessed by reflection somewhere and that's why this is needed?

elinor-fung replied:
Parts of it will be accessed by whatever COM server is firing the events for which a handler has been added. Specifically, it is the `ComEventsSink`'s implementation of `IDispatch` that is not explicitly used (directly or via reflection), but is necessary for handling COM events.

Is there a better way to prevent that from being trimmed out? I know there is `PreserveDependency`, but since we are not actually the ones that call into it, we don't actually have another method that is using it.

stephentoub replied:
Thanks for the explanation.  I think what you have is the best we can currently do.  That said, are you actually seeing the interface implementation get removed?  I didn't think the linker would do that if the interface is used anywhere, but my knowledge may be out of date.

elinor-fung replied:
Yeah, I hit this because the implementation of `IDispatch.Invoke` was getting removed.

jkotas replied:
https://github.com/mono/linker/issues/936 ?

VSadov commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs:
```diff
@@ -436,7 +442,20 @@ internal static object TryGetMemberValue(object obj, string name, bool ignoreExc
+#if ENABLECOMBINDER
+            else if (obj != null && ComObjectType.IsAssignableFrom(obj.GetType()))
+            {
+                var comExclusionList = new string[] { "MailEnvelope" }; //add any com names to be excluded from dynamic view here
```
>MailEnvelope [](start = 55, length = 12)

I wonder what is the history behind excluding `MailEnvelope` from the debug view and whether the exclusion is still relevant.

elinor-fung replied:
That's all there is in Framework as far as comments go. Looks like it has been there since the sources were added / copied in 2010 (not sure how to go back further).

I tried removing the exclusion. And was able to see the `MailEnvelope` property for an Excel `Worksheet` and a Word `Document` just fine.

jkotas commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs:
```diff
@@ -0,0 +1,578 @@
+
+        internal static MethodInfo GetGetIDispatchForObjectMethod()
+        {
+            // GetIDispatchForObject always throws a PNSE in .NET Core, so we work around it by using GetComInterfaceForObject with our IDispatch type.
```
Would it make sense to change `GetIDispatchForObject` to forward to `GetComInterfaceForObject` in .NET Core?

elinor-fung replied:
I think that makes sense. It looks like we also have `MarshalNative::GetIDispatchForObjectNative` (currently unused, I think).

I filed #33217 for it.

jkotas commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs:
```diff
@@ -0,0 +1,578 @@
+            Marshal.GetNativeVariantForObject(obj, ConvertVariantByrefToPtr(ref variant));
+        }
+
+        [Obsolete("do not use this method", true)]
```
These Obsolete attributes do not look useful. Are these methods accessed by reflection only? If yes, it should be enough to just add a comment. If not, delete them.

elinor-fung replied:
Yep - accessed via reflection (seems like adding an Obsolete attribute like this was common in framework for parts of System.Dynamic). Will remove and switch to a comment.

jkotas commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs:
```diff
@@ -0,0 +1,578 @@
+                    if (s_dynamicModule == null)
+                    {
+                        var attributes = new CustomAttributeBuilder[] {
+                            new CustomAttributeBuilder(typeof(UnverifiableCodeAttribute).GetConstructor(Array.Empty<Type>()), Array.Empty<object>())
```
UnverifiableCodeAttribute does nothing in .NET Core. No reason to emit it.

VSadov commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComBinderHelpers.cs:
```diff
@@ -0,0 +1,118 @@
+            return t.IsGenericType && t.GetGenericTypeDefinition() == typeof(StrongBox<>);
+        }
+
+        // this helper prepares arguments for COM binding by transforming ByVal StongBox arguments
```
>StongBox [](start = 80, length = 8)

StrongBox   (typo)

VSadov commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComEventSinksContainer.cs:
```diff
@@ -0,0 +1,75 @@
+
+        public static ComEventSinksContainer FromRuntimeCallableWrapper(object rcw, bool createIfNotFound)
+        {
+            // !!! Marshal.Get/SetComObjectData has a LinkDemand for UnmanagedCode which will turn into
```
The comment about LinkDemand and SecurityCritical can be removed now :-)

cston commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs:
```diff
@@ -220,6 +225,40 @@ public void RemoveDelegate(Delegate d)
+                foreach (int idx in toRemove)
+                {
+                    _delegateWrappers.RemoveAt(idx);
+                }
```
Do we need to remove higher indices before lower?

elinor-fung replied:
Ack - yes.

VSadov commented on src/libraries/Common/src/System/Runtime/InteropServices/ComEventsSink.cs:
```diff
@@ -10,7 +12,7 @@ namespace System.Runtime.InteropServices
     /// managed delegates to COM's connection point based events.
     /// </summary>
-    internal class ComEventsSink : IDispatch, ICustomQueryInterface
+    internal partial class ComEventsSink : IDispatch, ICustomQueryInterface
```
>IDispatch, ICustomQueryInterface [](start = 43, length = 32)

Nice! In olden time this had to use `IReflect/RealProxy` stuff to pretend it implements interfaces.

VSadov commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComObject.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            Debug.Assert(ComBinder.IsComObject(rcw));
+
+            // Marshal.Get/SetComObjectData has a LinkDemand for UnmanagedCode which will turn into
```
>// Marshal.Get/SetComObjectData has a LinkDeman [](start = 12, length = 47)

Another mention of LinkDemand

VSadov commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/Helpers.cs:
```diff
@@ -0,0 +1,52 @@
+
+    internal static class Requires
+    {
+        internal static void NotNull(object value, string paramName)
```
>NotNull [](start = 29, length = 7)

This made sense when this was public API. As long as the COM binder is internal, this could be `[Conditional("DEBUG")]`

VSadov commented on src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/Helpers.cs:
```diff
@@ -0,0 +1,52 @@
+            }
+        }
+
+        internal static void Condition(bool precondition, string paramName)
```
This made sense when this was public API. As long as the COM binder is internal, this could be `[Conditional("DEBUG")]`

cston commented on src/coreclr/tests/src/Interop/COM/Dynamic/Server/DispatchImpl.h:
```diff
@@ -0,0 +1,44 @@
+    void *_instance;
+};
+
+// Marco to use for defining ref counting impls
```
>Marco [](start = 3, length = 5)

Typo

cston commented on src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs:
```diff
@@ -0,0 +1,40 @@
+
+    class Program
+    {
+        static int Main(string[] doNotUse)
```
>Main [](start = 19, length = 4)

Are these tests run in CI builds?

elinor-fung replied:
Yes (as part of the CoreCLR tests)

elinor-fung commented on src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs:
```diff
@@ -0,0 +1,40 @@
+
+    /*
+       Param array : ParamArrayArgs
+     */
```
Bah - left over from notes to myself about tests to add. Will remove.

elinor-fung committed:
PR feedback

Fix delegate removal by index
Remove unnecessary comments and attributes

VSadov commented:
I left some comments, but there is nothing that is blocking. The change looks nice. 
- I like that the remoting/proxy pieces are no longer needed.
- It is nice to see that most IL-generated helpers (like ConvertByrefToPtr  flavors) are gone. Maybe one day we can get rid of the helpers that wrap `CALLI` as well.

LGTM

elinor-fung committed:
PR feedback

Remove exclusion list for dynamic member debug view
Simplify handling of ref enum parameter for events

jaredpar commented:
@VSadov 

> It is nice to see that most IL-generated helpers (like ConvertByrefToPtr flavors) are gone. Maybe one day we can get rid of the helpers that wrap CALLI as well.

Is this something that function pointers will help out with?  @333fred

VSadov commented:
@jaredpar - Yes, I think the whole point of function pointers is to make CALLI expressible in C# and handle scenarios like this.
